Add JS for simulation, tooltip, and halo

mortonanalytics opened the pull request:
This pull request addresses #27 except for labels (currently duplicates tool tip label). The review should **focus on `pySourceCredChart.js` and `pySourceCredChart.css`.** The other files are `nbextension` files that update as the library changes. Given my conversation with @decentralion, I'm going to pause here with it for a bit and see about importing React components - specifically, the Source Cred visualizer components.

The reason for this pull request is to make the subgraph inspection easier with a new simulation, a 'halo' on clicking a node, and a basic tool tip description of the node. The example notebook demonstrates the use for a smaller graph object - fyi @mzargham. Before, the chart worked well for the very large graph object but was not conducive to a more granular inspection.

I've also broken the code into various functions/components - though the underlying d3 hasn't changed for nodes and edges.

Commit:
add JS for simulation, halo, and tooltip

relates to #29

Commit:
correcting previous commit

Commit:
code cleanup

teamdandelion commented on infra/pySourceCredChart/js/lib/pySourceCredChart.js:
```diff
@@ -0,0 +1,320 @@
+var sourceCredChartView = widgets.DOMWidgetView.extend({
+	
+	render: function(){
+		var that = this;
```
style nit: personally, I prefer to use fat arrow functions `(arg) => { return result; }` rather than `function(arg) { return result; }` because fat arrows preserve `this` context and so tracking that is unnecessary. no change requested.

teamdandelion replied:
style nit: I use `const` as my default variable declaration, `let` for specific cases where I need to re-assign the variable, and `var` almost never.

teamdandelion commented on infra/pySourceCredChart/js/lib/pySourceCredChart.js:
```diff
@@ -0,0 +1,320 @@
+		this.simulation = _simulation(that, this.nodes, this.links);
+    
+		// update chart
+		// this.addZoom();
```
remove commented out code

teamdandelion commented on infra/pySourceCredChart/js/lib/pySourceCredChart.js:
```diff
@@ -0,0 +1,320 @@
+		this.nodes = this.data[this.mapping.nodes];
+		
+		/// define simulation object
+		this.simulation = _simulation(that, this.nodes, this.links);
```
i'm not fond of comments like "define the simulation object". it doesn't provide any context beyond the fact that the code defines an object called `this.simulation`. in other words, it provides no value now, but may become confusing later if the code changes so it goes out of date.

teamdandelion commented on infra/pySourceCredChart/js/lib/pySourceCredChart.js:
```diff
@@ -0,0 +1,320 @@
+		d3updateEdges(that, this.links);		
+		
+				
+		// Update and restart the simulation.
```
similarly, this comment seems unnecessary because it's clear from reading the code what is happening.

teamdandelion commented on infra/pySourceCredChart/js/lib/pySourceCredChart.js:
```diff
@@ -0,0 +1,320 @@
+		  setTimeout(function(){
+			console.log("call stop");
+			that.simulation.alpha(0).restart();
+			that.simulation.stop();
```
why do we set the alpha to 0, then restart it, and then stop it? wouldn't just stopping it be sufficient?

teamdandelion replied:
also, if you want the simulation to stop on its own over time, there's no need to do this with a timer. you can just set a target alpha lower than the minimum alpha, which is indeed the default settings for a force layout. the alpha values that I chose in the hackathon prototype were specifically chosen because I wanted the simulation to keep running forever.

cf [the docs on alpha decay](https://github.com/d3/d3-force#simulation_alphaDecay)

teamdandelion commented on infra/pySourceCredChart/js/lib/pySourceCredChart.js:
```diff
@@ -0,0 +1,320 @@
+		that.simulation.force("link").links(this.links);
+		that.simulation.alpha(1).restart();
+		
+		function timeIt(){
```
I'm curious why `timeIt` is declared as a function since it's always called exactly once and doesn't close over any variables. why not just elevate the setTimeout block directly into the containing function?

teamdandelion commented on infra/pySourceCredChart/js/lib/pySourceCredChart.js:
```diff
@@ -0,0 +1,320 @@
+			return 5
+			} 
+		
+		}
```
looks like the braces have become mismatched. i recommend running [prettier](https://prettier.io/) on this code, it will make it more consistent and legible.

teamdandelion commented on infra/pySourceCredChart/js/lib/pySourceCredChart.js:
```diff
@@ -0,0 +1,320 @@
+			// Use the square of the score as radius, so area will be proportional to score (if score available).
+			// For the Python implementation, the scores are separate from the node so we use the node index to
+			// find the score.
+			const _maxScore = 100;
```
why is maxScore 100? is this contract documented somewhere?
I would recommend just setting maxScore = d3.max(that.scores)

teamdandelion commented on infra/pySourceCredChart/js/lib/pySourceCredChart.js:
```diff
@@ -0,0 +1,320 @@
+	}
+	
+	/// d3 based DOM functions; trying to mimic React Components
+	function d3setScaffolding(that){
```
nit: i would rename it to `setupD3Scaffolding` (or at least `d3SetupScaffolding`)

teamdandelion commented on infra/pySourceCredChart/js/lib/pySourceCredChart.js:
```diff
@@ -0,0 +1,320 @@
+			.style('background', 'none repeat scroll 0 0 #ffffff');
+	}
+	
+	function d3updateNodes(that, nodes){
```
nit: i would rename it to `updateD3Nodes` (or at least `d3UpdateNodes`)

teamdandelion commented on infra/pySourceCredChart/js/lib/pySourceCredChart.js:
```diff
@@ -0,0 +1,320 @@
+		}
+	}
+	
+	function d3updateEdges(that, links){
```
same suggestion as with the other `d3...` method names

teamdandelion commented on infra/pySourceCredChart/js/lib/pySourceCredChart.js:
```diff
@@ -0,0 +1,320 @@
+		// edge data join
+		var edge = that.edge.selectAll('.edge').data(links);
+		
+		// edge exit
```
same suggestion as with the other comments that don't provide any context except for what's already in the code

mortonanalytics replied:
Oops, missed this one.

mortonanalytics replied:
Should be cleaned up now.

teamdandelion reviewed (commented):
Left a lot of style nits and a few substantive comments. :)

Commit:
code review changes

this commit reflects changes requested upon code review from @decentralion: variable declaration, using => for functions, removing unneccessary comments, and some other substantive changes. I also changed naming conventions from 'd3' to just the function name since I don't need to identify d3 DOM manipulation functions anymore. I also changed 'if' statements to switch calls for easier reading.

mortonanalytics commented:
@decentralion with commit [62e9802](https://github.com/sourcecred/research/pull/30/commits/62e980242cac560c9807af427900b1114803a52c ), I've made the requested and suggested changes. Thanks for the review!

Commit:
remove unhelpful comments

Commit:
Merge branch 'master' into rm2-exploratory

Commit:
remove 'rMorton' folder from this repo
